src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/JsonDictionaryConverter.cs

```diff
@@ -347,7 +355,9 @@ static TKey ReadDictionaryKey(JsonConverter<TKey> keyConverter, ref Utf8JsonRead
                     JsonSerializer.WriteMetadataForObject(this, ref state, writer);
                 }
 
-                state.Current.JsonPropertyInfo = state.Current.JsonTypeInfo.ElementTypeInfo!.PropertyInfoForTypeInfo;
+                state.Current.JsonPropertyInfo = jsonTypeInfo.ElementTypeInfo!.PropertyInfoForTypeInfo;
+
+                jsonTypeInfo.OnSerializing?.Invoke(dictionary);
```

Comment:
Likewise, I would invoke the callback before metadata or the start object token get written to the wire.

Reply:
Idem.